### PR TITLE
add connect overloads for lvalue-ref and const-lvalue-ref

### PR DESCRIPTION
**Pull request by kirkshoop**
This adds overloads of `connect() &` and `connect() const&` to a variety of senders.

This turns out to be a non-issue for tag_invoke because `Target&&` is a forwarding-reference while `connect() &&` does not provide access to a forwarding-reference to `this`.

### CODE REVIEW

File: include/unifex/stop_immediately.hpp [hunk 1]
```diff
@@ -296,6 +296,10 @@ struct _stream<SourceStream, Values...>::type {
     operation<Receiver> connect(Receiver&& receiver) && {
       return operation<Receiver>{stream_, (Receiver&&)receiver};
     }
+    template<typename Receiver>
+    void connect(Receiver&& receiver) & =delete;
```
Comment (lewissbaker) [inline include/unifex/stop_immediately.hpp]: Is this overload declaration necessary given that we have the `const &` overload below?

File: include/unifex/stop_immediately.hpp [hunk 2]
```diff
@@ -413,6 +417,14 @@ struct _stream<SourceStream, Values...>::type {
...
       return operation<Receiver>{stream_, (Receiver &&) receiver};
     }
+    template<typename Receiver>
+    operation<Receiver> connect(Receiver&& receiver) & {
+      return operation<Receiver>{stream_, (Receiver &&) receiver};
+    }
+    template<typename Receiver>
+    operation<Receiver> connect(Receiver&& receiver) const& {
+      return operation<Receiver>{stream_, (Receiver &&) receiver};
+    }
```
Comment (lewissbaker) [inline include/unifex/stop_immediately.hpp]: I'm pretty sure it doesn't make sense to allow repeatedly running the cleanup operation here.
This cleanup operation implementation shared references to the stream object and assumes that it will only be run once - the state-machine does not allow for running the cleanup operations multiple times.

In general, I think cleanup operations should generally not be repeatable. They will often have side-effects that release resources and should only be run once.

It also seems suspicious that each of these are performing the same

File: include/unifex/sync_wait.hpp [hunk 3]
```diff
@@ -188,7 +188,7 @@ namespace _sync_wait_cpo {
 
         auto operation = connect(
           (Sender&&)sender,
-          _sync_wait::thread_unsafe_receiver<Result, StopToken&&>{
+          _sync_wait::thread_unsafe_receiver<Result, StopToken>{
```
Comment (lewissbaker) [inline include/unifex/sync_wait.hpp]: Why was this change necessary?

File: include/unifex/thread_unsafe_event_loop.hpp [hunk 4]
```diff
@@ -149,6 +149,16 @@ namespace _thread_unsafe_event_loop {
...
+    template <typename Receiver>
+    after_operation<Duration, std::remove_cvref_t<Receiver>> connect(Receiver&& r) & {
+      return after_operation<Duration, std::remove_cvref_t<Receiver>>{
+          (Receiver &&) r, duration_, loop_};
+    }
+    template <typename Receiver>
+    after_operation<Duration, std::remove_cvref_t<Receiver>> connect(Receiver&& r) const& {
+      return after_operation<Duration, std::remove_cvref_t<Receiver>>{
+          (Receiver &&) r, duration_, loop_};
+    }
```
Comment (lewissbaker) [inline include/unifex/thread_unsafe_event_loop.hpp]: Can we combine these three overloads into a single `const&` overload to avoid duplication here?

File: include/unifex/thread_unsafe_event_loop.hpp [hunk 5]
```diff
@@ -225,6 +235,16 @@ namespace _thread_unsafe_event_loop {
...
+    template <typename Receiver>
+    at_operation<std::remove_cvref_t<Receiver>> connect(Receiver&& r) & {
+      return at_operation<std::remove_cvref_t<Receiver>>{
+          (Receiver &&) r, dueTime_, loop_};
+    }
+    template <typename Receiver>
+    at_operation<std::remove_cvref_t<Receiver>> connect(Receiver&& r) const& {
+      return at_operation<std::remove_cvref_t<Receiver>>{
+          (Receiver &&) r, dueTime_, loop_};
+    }
```
Comment (lewissbaker) [inline include/unifex/thread_unsafe_event_loop.hpp]: Can we combine these overloads into a single `const&` overload to avoid the duplication?

File: include/unifex/timed_single_thread_context.hpp [hunk 6]
```diff
@@ -134,7 +134,17 @@ namespace _timed_single_thread_context {
...
+    template <typename Receiver>
+    after_operation<Duration, Receiver> connect(Receiver&& receiver) & {
+      return after_operation<Duration, Receiver>{
+          *context_, duration_, (Receiver &&) receiver};
+    }
+    template <typename Receiver>
+    after_operation<Duration, Receiver> connect(Receiver&& receiver) const & {
       return after_operation<Duration, Receiver>{
           *context_, duration_, (Receiver &&) receiver};
     }
```
Comment (lewissbaker) [inline include/unifex/timed_single_thread_context.hpp]: Can we combine these overloads into a single `const&` overload to avoid the duplication?

File: include/unifex/trampoline_scheduler.hpp [hunk 7]
```diff
@@ -123,6 +123,14 @@ class scheduler {
...
       return operation<Receiver>{(Receiver &&) receiver, maxRecursionDepth_};
     }
+    template <typename Receiver>
+    operation<Receiver> connect(Receiver&& receiver) & {
+      return operation<Receiver>{(Receiver &&) receiver, maxRecursionDepth_};
+    }
+    template <typename Receiver>
+    operation<Receiver> connect(Receiver&& receiver) const& {
+      return operation<Receiver>{(Receiver &&) receiver, maxRecursionDepth_};
+    }
```
Comment (lewissbaker) [inline include/unifex/trampoline_scheduler.hpp]: Combine into a single `const&` overload?

File: include/unifex/transform.hpp [hunk 8]
```diff
@@ -170,6 +171,7 @@ struct _sender<Predecessor, Func>::type {
 
   template <typename Receiver>
   auto connect(Receiver&& r) &
+      noexcept(is_nothrow_tag_invocable_v<tag_t<unifex::connect>, Predecessor&, receiver<std::remove_cvref_t<Receiver>>>)
```
Comment (lewissbaker) [inline include/unifex/transform.hpp]: Also needs to check that both `Receiver` and `Func` is nothrow-decay-copyable (as a proxy for detecting that `receiver` is nothrow constructible.

Similarly for the other overloads.

File: include/unifex/transform_stream.hpp [hunk 9]
```diff
@@ -27,7 +27,7 @@ namespace _tfx_stream {
     auto operator()(StreamSender&& stream, Func&& func) const {
       return next_adapt_stream(
           (StreamSender &&) stream, [func = (Func &&) func](auto&& sender) mutable {
-            return transform((decltype(sender))sender, std::ref(func));
+            return transform((decltype(sender))sender, (Func &&)func);
```
Comment (lewissbaker) [inline include/unifex/transform_stream.hpp]: The use of `std::ref()` here was intentional to allow the use of stateful function objects.

Also, the function will be called once for each `next()` and so we shouldn't be moving the function into the transform-sender.

File: include/unifex/when_all.hpp [hunk 10]
```diff
@@ -63,8 +63,8 @@ template <
     typename... Rest>
 struct _operation_tuple<Index, Receiver, First, Rest...>::type
   : operation_tuple<Index + 1, Receiver, Rest...> {
-  template <typename Parent>
-  explicit type(Parent& parent, First&& first, Rest&&... rest)
+  template <typename First2, typename Parent>
+  explicit type(Parent& parent, First2&& first, Rest&&... rest)
```
Comment (lewissbaker) [inline include/unifex/when_all.hpp]: Why was this change necessary?
Shouldn't the sender be perfectly forwarded through to this constructor?

File: include/unifex/when_all.hpp [hunk 11]
```diff
@@ -194,9 +194,10 @@ struct _op<Receiver, Senders...>::type {
   template<std::size_t Index, typename Operation>
   friend class _element_receiver;
 
-  explicit type(Receiver&& receiver, Senders&&... senders)
+  template<typename... Senders2>
+  explicit type(Receiver&& receiver, Senders2&&... senders)
     : receiver_((Receiver &&) receiver),
-      ops_(*this, (Senders &&) senders...) {}
+      ops_(*this, (Senders2 &&) senders...) {}
```
Comment (lewissbaker) [inline include/unifex/when_all.hpp]: The senders should be perfectly forwarded through here so I wouldn't have thought that the Senders2 template parameter would have been necessary.
What problems were you seeing with this?

File: include/unifex/finally.hpp [hunk 12]
```diff
@@ -700,39 +701,40 @@ namespace unifex
...
                 is_connectable<
                   CompletionSender,
                   done_receiver<
                       SourceSender,
                       CompletionSender,
-                      Receiver>>>,
-              int> = 0>
+                      std::remove_cvref_t<Receiver>>>,
+                std::true_type>,
+              int(*)[__LINE__]> = nullptr>
```
[the remark this answers is not in the document]
Comment (kirkshoop) [reply include/unifex/finally.hpp]: yes.. though the result is a much longer symbol :(

File: include/unifex/when_all.hpp (hunk 11 above)
Comment (kirkshoop) [reply include/unifex/when_all.hpp]: I reverted this and it compiles locally. Thanks!

File: include/unifex/when_all.hpp (hunk 10 above)
Comment (kirkshoop) [reply include/unifex/when_all.hpp]: I reverted this and it compiles locally. Thanks!

File: include/unifex/transform_stream.hpp (hunk 9 above)
Comment (kirkshoop) [reply include/unifex/transform_stream.hpp]: this appears to work now - I have reverted the change

File: include/unifex/sync_wait.hpp (hunk 3 above)
Comment (kirkshoop) [reply include/unifex/sync_wait.hpp]: && did not look like it was doing anything.

File: include/unifex/stop_immediately.hpp (hunk 2 above)
Comment (kirkshoop) [reply include/unifex/stop_immediately.hpp]: Destructors should only be run once and should be run on a const object.

I do not think that const connect maps directly to repeatable.

File: include/unifex/stop_immediately.hpp (hunk 1 above)
Comment (kirkshoop) [reply include/unifex/stop_immediately.hpp]: I wanted to be explicit that this was removed on purpose. could be a comment..

File: include/unifex/sequence.hpp [hunk 13]
```diff
@@ -288,7 +288,7 @@ namespace unifex
           Successor,
           Receiver>;
 
-      Successor successor_;
+      std::remove_cvref_t<Successor> successor_;
```
[the remark this answers is not in the document]
Comment (kirkshoop) [reply include/unifex/sequence.hpp]: this was an incomplete rollback of an intermediate state. removed this and others.

File: include/unifex/range_stream.hpp [hunk 14]
```diff
@@ -63,6 +63,12 @@ struct next_sender {
   operation<Receiver> connect(Receiver&& receiver) && {
     return operation<Receiver>{stream_, (Receiver &&) receiver};
   }
+  template <typename Receiver>
+  operation<Receiver> connect(Receiver&& receiver) & {
+    return operation<Receiver>{stream_, (Receiver &&) receiver};
+  }
+  template <typename Receiver>
+  void connect(Receiver&& receiver) const& = delete;
```
[the remark this answers is not in the document]
Comment (kirkshoop) [reply include/unifex/range_stream.hpp]: I marked the lvalue connect delete as well.

File: include/unifex/stop_immediately.hpp (hunk 2 above)
Comment (lewissbaker) [reply include/unifex/stop_immediately.hpp]: > Destructors should only be run once and should be run on a const object.

Are you equating the `cleanup()` operation to a destructor here?

I guess for ordinary types, you can call the destructor on a const object, however the object ceases to be a const object inside the destructor (I still want to be able to mutate data members inside a destructor). Also, a destructor consumes an object.

So I can perhaps see an argument for allowing the `cleanup()` method to be declared const-rvalue. However, there isn't an easy way to declare that the method was const-rvalue callable but that inside the method the data-members were all mutable without sprinkling const_cast everywhere.

I think the qualification on the `cleanup()` method is a separate concern from the qualification of the `connect()` method on the sender returned from `cleanup()`.

I don't want people to think that it's safe to call `connect()` multiple times on this sender. The current way we indicate that it is safe to call `connect()` multiple times is by adding a const-lvalue overload of `connect()`.

File: include/unifex/sync_wait.hpp (hunk 3 above)
Comment (lewissbaker) [reply include/unifex/sync_wait.hpp]: It was avoiding a copy of the StopToken and instead capturing a reference to the stop-token passed to sync_wait.

With this change it will still capture a reference to the stop-token in the receiver if the caller passed an lvalue, but will now move-construct the stop-token into the receiver if the caller passed an rvalue.

If we think that we should be capturing a copy of the stop-token in the receiver instead of a reference then perhaps we should be decay-copying the stop-token here instead?

I guess this is all kind of moot after PR #80 is merged anyway.

File: examples/async_trace.cpp [hunk 15]
```diff
@@ -74,26 +80,28 @@ int main() {
 
   auto start = steady_clock::now();
 
-  sync_wait(transform(
-      when_all(
+  sync_wait(
+    transform(
+      defer([&]{
```
Comment (lewissbaker) [inline examples/async_trace.cpp]: Why was defer() required here?
There shouldn't be a need for repeatability of the operation if just passing to sync_wait() without a repeat/retry of some sort.

File: include/unifex/async_trace.hpp [hunk 16]
```diff
@@ -199,7 +199,11 @@ namespace _async_trace {
     using error_types = Variant<std::exception_ptr>;
 
     template <typename Receiver>
-    operation<Receiver> connect(Receiver&& r) {
+    operation<Receiver> connect(Receiver&& r) && {
+      return operation<Receiver>{(Receiver &&) r};
+    }
```
Comment (lewissbaker) [inline include/unifex/async_trace.hpp]: Remove this overload? It's doing the same thing as the const& version below.

File: include/unifex/finally.hpp [hunk 17]
```diff
@@ -590,16 +590,17 @@ namespace unifex
         manual_lifetime_union<std::exception_ptr, Errors...>;
 
     public:
-      template <typename CompletionSender2, typename Receiver2>
+      type() = default;
```
Comment (lewissbaker) [inline include/unifex/finally.hpp]: Why was it necessary to declare a default constructor here?
I wouldn't expect it to ever be called.

File: include/unifex/finally.hpp [hunk 18]
```diff
@@ -590,16 +590,17 @@ namespace unifex
...
+      template <typename SourceSender2, typename CompletionSender2, typename Receiver2>
       explicit type(
-          SourceSender&& sourceSender,
+          SourceSender2&& sourceSender,
```
[the remark this answers is not in the document]
Comment (lewissbaker) [reply include/unifex/finally.hpp]: Not fixed yet?

File: include/unifex/finally.hpp [hunk 19]
```diff
@@ -617,19 +618,19 @@ namespace unifex
...
 
     private:
-      UNIFEX_NO_UNIQUE_ADDRESS CompletionSender completionSender_;
+      UNIFEX_NO_UNIQUE_ADDRESS std::remove_cvref_t<CompletionSender> completionSender_;
       UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
       bool started_ = false;
 
       // Result storage.
       union {
         // Storage for error-types that might be produced by ValueSender.
```
Comment (lewissbaker) [inline include/unifex/finally.hpp]: Just noticed that comment should probably be updated to reference SourceSender instead of ValueSender.

File: include/unifex/finally.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (lewissbaker) [reply include/unifex/finally.hpp]: On the other hand, using __LINE__ means that adding a comment anywhere in the file above can change the ABI of this function.

File: include/unifex/finally.hpp [hunk 20]
```diff
@@ -700,39 +701,41 @@ namespace unifex
...
+                      std::remove_cvref_t<Receiver>>>,
                 is_connectable<
                   CompletionSender,
                   done_receiver<
                       SourceSender,
                       CompletionSender,
-                      Receiver>>>,
-              int> = 0>
+                      std::remove_cvref_t<Receiver>>>,
+                std::true_type>,
```
Comment (lewissbaker) [inline include/unifex/finally.hpp]: Why was it necessary to add `std::true_type` here?
I thought std::conjunction would have had the same behaviour without it?

File: include/unifex/finally.hpp [hunk 21]
```diff
@@ -700,39 +701,41 @@ namespace unifex
...
       template <
           typename Receiver,
           typename CPO,
           typename S,
           std::enable_if_t<
               std::conjunction_v<
                 std::is_same<CPO, tag_t<connect>>,
-                std::is_same<S, sender>,
+                std::is_same<std::remove_cvref_t<S>, sender>,
+                std::negation<std::is_lvalue_reference<S>>,
```
Comment (lewissbaker) [inline include/unifex/finally.hpp]: Was it necessary to limit to rvalue-reference only here so that the subsequent `is_connectable<SourceSender/CompletionSender` checks are using the correct value-category?

ie. they'd need to be checking `is_connectable<const SourceSender&, ...>` and also `is_copy_constructible<CompletionSender>` and `is_connectable<CompletionSender, ...>`.

Do we still need to add this other const-lvalue connect overload?

File: include/unifex/just.hpp [hunk 22]
```diff
@@ -72,26 +72,38 @@ class _sender<Values...>::type {
...
 
-  template <typename Receiver>
+  template <
+    typename Value2,
+    std::enable_if_t<
+      (sizeof...(Values) == 1 &&
+      !std::is_same_v<Value2, type>), int> = 0>
+    explicit type(Value2&& value)
+    noexcept(std::is_nothrow_constructible_v<std::tuple<Values...>, Value2>)
+    : values_((Value2 &&) value) {}
```
Comment (lewissbaker) [inline include/unifex/just.hpp]: I wonder whether this could all be simplified by adding an extra std::in_place_t arg in front of the parameter pack to disambiguate the in-place constructors from the copy/move constructors?

e.g.
```
template<typename... Values2>
explicit type(std::in_place_t, Values2&&... values)
  noexcept(std::is_nothrow_constructible_v<std::tuple<Values...>, Values2...>)
  : values_((Values2&&)values...)
{}
```

Then just update the `just()` implementation to pass `std::in_place` as the first arg to the constructor.

File: include/unifex/manual_event_loop.hpp [hunk 23]
```diff
@@ -93,6 +93,10 @@ class context {
       operation<Receiver> connect(Receiver&& receiver) && {
         return operation<Receiver>{(Receiver &&) receiver, loop_};
       }
```
Comment (lewissbaker) [inline include/unifex/manual_event_loop.hpp]: Can we remove the rvalue overload now?

File: include/unifex/ready_done_sender.hpp [hunk 24]
```diff
@@ -55,7 +55,11 @@ namespace _ready_done {
     }
 
     template <typename Receiver>
-    operation<Receiver> connect(Receiver&& receiver) {
+    operation<Receiver> connect(Receiver&& receiver) && {
+      return operation<Receiver>{(Receiver &&) receiver};
+    }
```
Comment (lewissbaker) [inline include/unifex/ready_done_sender.hpp]: Can we remove the rvalue overload now?

File: include/unifex/retry_when.hpp [hunk 25]
```diff
@@ -352,28 +352,14 @@ class _sender<Source, Func>::type {
...
     typename Receiver,
     std::enable_if_t<
         std::is_constructible_v<Source, const Source&> &&
         std::is_constructible_v<Func, const Func&> &&
         std::is_constructible_v<std::remove_cvref_t<Receiver>, Receiver> &&
         is_connectable_v<Source&, source_receiver<Source, Func, std::remove_cvref_t<Receiver>>>, int> = 0>
-  operation<Source, Func, Receiver> connect(Receiver&& r) &
+  operation<Source, Func, Receiver> connect(Receiver&& r) const&
       noexcept(std::is_nothrow_constructible_v<
           operation<Source, Func, Receiver>, const Source&, const Func&, Receiver>) {
```
Comment (lewissbaker) [inline include/unifex/retry_when.hpp]: I wonder if the noexcept clause should be defined in terms of nothrow copyability of Source, Func, Receiver rather than of nothrow constructibility of operation to avoid needing to instantiate the operation class too early.

File: include/unifex/sequence.hpp [hunk 26]
```diff
@@ -298,7 +298,7 @@ namespace unifex
       status status_;
       union {
         manual_lifetime<operation_t<
-            Predecessor,
+            std::remove_cvref_t<Predecessor>,
```
Comment (lewissbaker) [inline include/unifex/sequence.hpp]: This should just be `Predecessor` rather than `std::remove_cvref_t<Predecessor>` to match the type we are passing to `connect()` in the constructor.

File: include/unifex/stop_immediately.hpp [hunk 27]
```diff
@@ -413,6 +415,10 @@ struct _stream<SourceStream, Values...>::type {
     operation<Receiver> connect(Receiver&& receiver) && {
       return operation<Receiver>{stream_, (Receiver &&) receiver};
     }
+    template<typename Receiver>
+    operation<Receiver> connect(Receiver&& receiver) const& {
+      return operation<Receiver>{stream_, (Receiver &&) receiver};
+    }
```
Comment (lewissbaker) [inline include/unifex/stop_immediately.hpp]: The cleanup_sender should not be const-lvalue connectable.
The cleanup operation has side-effects that are not idempotent and starting it multiple times would have undefined-behaviour.

File: include/unifex/thread_unsafe_event_loop.hpp (hunk 4 above)
Comment (lewissbaker) [reply include/unifex/thread_unsafe_event_loop.hpp]: Can we remove the rvalue overload now?

File: include/unifex/thread_unsafe_event_loop.hpp [hunk 28]
```diff
@@ -225,6 +230,11 @@ namespace _thread_unsafe_event_loop {
       return at_operation<std::remove_cvref_t<Receiver>>{
           (Receiver &&) r, dueTime_, loop_};
     }
+    template <typename Receiver>
+    at_operation<std::remove_cvref_t<Receiver>> connect(Receiver&& r) const& {
```
Comment (lewissbaker) [inline include/unifex/thread_unsafe_event_loop.hpp]: Can we remove the rvalue connect() overload now that we have a const-lvalue overload?

File: include/unifex/timed_single_thread_context.hpp [hunk 29]
```diff
@@ -134,7 +134,12 @@ namespace _timed_single_thread_context {
     using error_types = Variant<>;
 
     template <typename Receiver>
-    after_operation<Duration, Receiver> connect(Receiver&& receiver) {
+    after_operation<Duration, Receiver> connect(Receiver&& receiver) && {
+      return after_operation<Duration, Receiver>{
+          *context_, duration_, (Receiver &&) receiver};
+    }
```
Comment (lewissbaker) [inline include/unifex/timed_single_thread_context.hpp]: Can we remove the rvalue overload now?

File: include/unifex/trampoline_scheduler.hpp [hunk 30]
```diff
@@ -123,6 +123,10 @@ class scheduler {
     operation<Receiver> connect(Receiver&& receiver) && {
       return operation<Receiver>{(Receiver &&) receiver, maxRecursionDepth_};
     }
```
Comment (lewissbaker) [inline include/unifex/trampoline_scheduler.hpp]: Can we remove the rvalue overload now?

File: include/unifex/transform.hpp [hunk 31]
```diff
@@ -161,23 +161,23 @@ struct _sender<Predecessor, Func>::type {
 
   template <typename Receiver>
   auto connect(Receiver&& r) &&
+      noexcept(
+        std::is_nothrow_constructible_v<std::remove_cvref_t<Receiver>, Receiver> && 
+        std::is_nothrow_copy_constructible_v<Func> &&
+        is_nothrow_tag_invocable_v<tag_t<unifex::connect>, Predecessor, receiver<std::remove_cvref_t<Receiver>>>)
```
Comment (lewissbaker) [inline include/unifex/transform.hpp]: Simplify using `is_nothrow_connectable_v`

File: include/unifex/transform.hpp [hunk 32]
```diff
@@ -161,23 +161,23 @@ struct _sender<Predecessor, Func>::type {
 
   template <typename Receiver>
   auto connect(Receiver&& r) &&
+      noexcept(
+        std::is_nothrow_constructible_v<std::remove_cvref_t<Receiver>, Receiver> && 
+        std::is_nothrow_copy_constructible_v<Func> &&
```
Comment (lewissbaker) [inline include/unifex/transform.hpp]: Should be `std::is_nothrow_move_constructible_v<Func>`

File: include/unifex/transform.hpp [hunk 33]
```diff
@@ -161,23 +161,23 @@ struct _sender<Predecessor, Func>::type {
...
-        pred_,
-        receiver<std::remove_cvref_t<Receiver>>{func_, std::forward<Receiver>(r)});
-  }
-
   template <typename Receiver>
   auto connect(Receiver&& r) const &
+      noexcept(
+        std::is_nothrow_constructible_v<std::remove_cvref_t<Receiver>, Receiver> && 
+        std::is_nothrow_copy_constructible_v<Func> &&
+        is_nothrow_tag_invocable_v<tag_t<unifex::connect>, const Predecessor&, receiver<std::remove_cvref_t<Receiver>>>)
```
Comment (lewissbaker) [inline include/unifex/transform.hpp]: Can be simplified using `is_nothrow_connectable_v`

File: include/unifex/finally.hpp (hunk 17 above)
Comment (kirkshoop) [reply include/unifex/finally.hpp:593]: at one point the variadic constructor was eating the default constructor call as well. I am using the std::in_place suggestion now.

File: examples/async_trace.cpp (hunk 15 above)
Comment (kirkshoop) [reply examples/async_trace.cpp:85]: I narrowed down the scope at which to put this `defer()` to get rid of a compile error.

File: include/unifex/finally.hpp (hunk 20 above)
Comment (kirkshoop) [reply include/unifex/finally.hpp:730]: It allowed me to comment out full lines easily. removed.

File: include/unifex/finally.hpp (hunk 21 above)
Comment (kirkshoop) [reply include/unifex/finally.hpp:717]: I cannot specify an rvalue overload of tag_invoke. it always creates a forwarding overload of tag_invoke when using &&. I just decided to use the same forwarding overload to handle both rvalue and const lvalue.

Comment (lewissbaker) [reply include/unifex/finally.hpp:716]: `std::is_lvalue_reference<S>` will still evaluate to true if `S` is a const-ref.
So this constraint would seem to only match rvalue-references as it stands.

Consider changing to `std::negation<std::is_same<S, sender&>>` if you want it to match `const sender&` and `sender&&`.
But if you do that then you also need to change the `is_connectable` check below to use the corresponding value category for the SourceSender check.

File: include/unifex/finally.hpp [hunk 34]
```diff
@@ -700,39 +700,40 @@ namespace unifex
...
       template <
           typename Receiver,
           typename CPO,
           typename S,
           std::enable_if_t<
               std::conjunction_v<
                 std::is_same<CPO, tag_t<connect>>,
-                std::is_same<S, sender>,
+                std::is_same<std::remove_cvref_t<S>, sender>,
+                std::negation<std::is_same<S, sender&>>,
```
[the remark this answers is not in the document]
Comment (kirkshoop) [reply include/unifex/finally.hpp:716]: for now I omitted the constraint

File: include/unifex/finally.hpp [hunk 35]
```diff
@@ -713,7 +713,7 @@ namespace unifex
               std::conjunction_v<
                 std::is_same<CPO, tag_t<connect>>,
                 std::is_same<std::remove_cvref_t<S>, sender>,
-                std::negation<std::is_lvalue_reference<S>>,
+                std::negation<std::is_same<S, sender&>>,
                 is_connectable<
                   SourceSender,
```
[the remark this answers is not in the document]
Comment (kirkshoop) [reply include/unifex/finally.hpp:718]: I used this to recover the value category for SourceSender
  `typename SourceSenderType = decltype(std::declval<S>().source_),`

File: include/unifex/retry_when.hpp [hunk 36]
```diff
@@ -346,36 +346,26 @@ class _sender<Source, Func>::type {
...
     return operation<Source, Func, Receiver>{
         (Source&&)source_, (Func&&)func_, (Receiver&&)r};
   }
 
   template<
     typename Receiver,
-    typename Op = operation<Source, Func, Receiver>,
     std::enable_if_t<
         std::is_constructible_v<Source, Source&> &&
         std::is_constructible_v<Func, Func&> &&
```
Comment (lewissbaker) [inline include/unifex/retry_when.hpp:361]: The `Source` and `Func` constraints should be using `is_copy_constructible_v`.

File: include/unifex/retry_when.hpp [hunk 37]
```diff
@@ -346,36 +346,26 @@ class _sender<Source, Func>::type {
         std::is_constructible_v<std::remove_cvref_t<Receiver>, Receiver> &&
         is_connectable_v<Source&, source_receiver<Source, Func, std::remove_cvref_t<Receiver>>>, int> = 0>
   operation<Source, Func, Receiver> connect(Receiver&& r) &&
-      noexcept(std::is_nothrow_constructible_v<
-          operation<Source, Func, Receiver>, Source, Func, Receiver>) {
+      noexcept(
+        std::is_nothrow_move_constructible_v<Source> &&
+        std::is_nothrow_move_constructible_v<Func> &&
+        std::is_nothrow_constructible_v<std::remove_cvref_t<Receiver>, Receiver>) {
```
Comment (lewissbaker) [inline include/unifex/retry_when.hpp:352]: Needs to also check `is_nothrow_connectable_v<Source&, source_receiver<Source, Func, std::remove_cvref_t<Receiver>>>` in the `noexcept` clause.

Comment (lewissbaker) [reply include/unifex/retry_when.hpp:352]: And also in the `connect() const &`  noexcept clause below.

File: examples/async_trace.cpp (hunk 15 above)
Comment (lewissbaker) [reply examples/async_trace.cpp:85]: I think I found the cause of this issue.

The transform() sender's connect() methods were not SFINAE-friendly and so when trying to call connect() on an rvalue it was still trying to instantiate the const-lvalue-ref connect() overload. Once transform() and when_all()'s connect() methods were modified to SFINAE out if input senders were not connectable then this defer() is no longer required.

See #106
